keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
+topology and zone-level topology, which is not necessarily terrible;
+the zone-level topology would only be useful when the node-level
+topology "failed". But if we wanted to cover that case, then we could
+have a separate annotation for node-level topology.)
+
+#### Approach 2: Automatic
+
+Alternatively, the EndpointSlice controller could attempt to determine
+automatically which services would benefit from node-level topology.
+For example, the EndpointSlice controller could look for services
+where:
+
+  - The service has an endpoint on every node, or at least "almost
+    every" node. (eg, no more than N or N% of nodes are missing an
+    endpoint).
+
+  - The pods that make up the service endpoints all have an
+    `OwnerReference` pointing to the same `DaemonSet`.
+
+  - ...
+
+If all relevant criteria are met, then it could set a new field in the
+`Hints` of the EndpointSlices for the service, and kube-proxy would
+use this to know to do node-level topology.
+
+#### Approach 3: Semi-Automatic or Automatic-With-Opt-Out
+
+Some combination of the above two approaches, where the EndpointSlice
+controller requires that the `topology-aware-hints` annotation (or
+another annotation) either is, or is not, set on the Service, before
+investigating the endpoints to decide whether to enable the feature.
+
+("Automatic with opt-out" is probably not a good idea; it implies that
+there are some services where users wouldn't want node-level topology,
+but also implies that we will break those services on upgrade for
+users who forgot to pre-emptively opt out when upgrading to the
+version of Kubernetes where this feature is enabled by default.)
+
+### Test Plan
```

Comment:
Please include recently added updates to the [Test Plan section of the KEP template] (https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template#test-plan) here.
These sections are required in order for the enhancement to be included in releases.